File: src/GitRemoteS3/Lfs.hs
-- SPDX-FileCopyrightText: 2023-present Amazon.com, Inc. or its affiliates
--
-- SPDX-License-Identifier: Apache-2.0

-- | Git LFS custom-transfer agent implementation.
--
-- Mirrors @git_remote_s3/lfs.py@.
--
-- Protocol: the agent reads JSON lines from stdin and writes JSON
-- lines to stdout.  See the git-lfs custom-transfer documentation:
-- https://github.com/git-lfs/git-lfs/blob/main/docs/custom-transfers.md
module GitRemoteS3.Lfs
  ( LFSProcess (..)
  , LFSS3Ops (..)
  , newLFSProcess
  , lfsUpload
  , lfsDownload
  , lfsInstall
  , lfsMain
  ) where

import           Control.Concurrent         (MVar, newMVar, withMVar, modifyMVar_)
import           Control.Exception          (SomeException, try)
import           Control.Monad              (unless, when)
import           Data.Aeson                 ( Value (..), object, (.=)
                                            , encode )
import qualified Data.Aeson                as Aeson
import qualified Data.Aeson.KeyMap         as KM
import qualified Data.Aeson.Key            as AKey
import qualified Data.ByteString.Lazy.Char8 as BLC
import           Data.IORef                 (IORef, newIORef, readIORef, modifyIORef')
import           Data.Maybe                 (fromMaybe)
import qualified Data.Text                 as T
import           System.Directory           (makeAbsolute, createDirectoryIfMissing)
import           System.Exit                (exitWith, exitSuccess, ExitCode (..))
import           System.IO                  ( hPutStrLn, hFlush, stderr, stdout
                                            , hSetBuffering, BufferMode (..)
                                            , stdin )
import           System.Process             (readProcessWithExitCode)

import           GitRemoteS3.Common         (parseGitUrl, ParseResult (..))
import           GitRemoteS3.Git            (validateRefName)

-- ---------------------------------------------------------------------------
-- Progress callback
-- ---------------------------------------------------------------------------

-- | Emits @progress@ JSON events to stdout as bytes are transferred.
-- Mirrors the Python @ProgressPercentage@ callable class.
data ProgressCallback = ProgressCallback
  { pcOid        :: String
  , pcSeenSoFar  :: IORef Integer
  , pcLock       :: MVar ()
  }

newProgressCallback :: String -> IO ProgressCallback
newProgressCallback oid = do
  seenRef <- newIORef 0
  lock    <- newMVar ()
  return ProgressCallback
    { pcOid       = oid
    , pcSeenSoFar = seenRef
    , pcLock      = lock
    }

reportProgress :: ProgressCallback -> Integer -> IO ()
reportProgress cb bytesAmount =
  withMVar (pcLock cb) $ \_ -> do
    modifyIORef' (pcSeenSoFar cb) (+ bytesAmount)
    seenSoFar <- readIORef (pcSeenSoFar cb)
    let evt = object
          [ "event"          .= ("progress" :: String)
          , "oid"            .= pcOid cb
          , "bytesSoFar"     .= seenSoFar
          , "bytesSinceLast" .= bytesAmount
          ]
    BLC.putStrLn (encode evt)
    hFlush stdout

-- ---------------------------------------------------------------------------
-- Error event helper
-- ---------------------------------------------------------------------------

writeErrorEvent :: String -> String -> IO ()
writeErrorEvent oid errMsg = do
  let evt = object
        [ "event" .= ("complete" :: String)
        , "oid"   .= oid
        , "error" .= object [ "code" .= (2 :: Int), "message" .= errMsg ]
        ]
  BLC.putStrLn (encode evt)
  hFlush stdout

-- ---------------------------------------------------------------------------
-- LFSProcess record
-- ---------------------------------------------------------------------------

-- | State for a running LFS transfer session.
data LFSProcess = LFSProcess
  { lfsBucket  :: String
  , lfsPrefix  :: String
  , lfsProfile :: Maybe String
  , lfsS3Ops   :: LFSS3Ops   -- ^ injectable S3 operations
  }

-- | Minimal S3 operations required by the LFS agent.
data LFSS3Ops = LFSS3Ops
  { lfsS3ObjectExists   :: String         -- ^ key
                        -> IO Bool
  , lfsS3UploadFile     :: FilePath       -- ^ local path
                        -> String         -- ^ key
                        -> (Integer -> IO ())  -- ^ progress callback
                        -> IO ()
  , lfsS3DownloadFile   :: String         -- ^ key
                        -> FilePath       -- ^ destination
                        -> (Integer -> IO ())  -- ^ progress callback
                        -> IO ()
  }

-- | Build an 'LFSProcess' from a parsed S3 URI.
-- Returns @Nothing@ when the URI is invalid (writes an error JSON
-- event to stdout and returns @Nothing@).
newLFSProcess :: LFSS3Ops -> String -> IO (Maybe LFSProcess)
newLFSProcess ops s3uri =
  case parseGitUrl s3uri of
    Nothing -> do
      let evt = object
            [ "error" .= object
                [ "code"    .= (32 :: Int)
                , "message" .= ("s3 uri " ++ s3uri ++ " is invalid")
                ]
            ]
      BLC.putStrLn (encode evt)
      hFlush stdout
      return Nothing
    Just (ParseResult _ _ _ Nothing) -> do
      let evt = object
            [ "error" .= object
                [ "code"    .= (32 :: Int)
                , "message" .= ("s3 uri " ++ s3uri ++ " is missing prefix")
                ]
            ]
      BLC.putStrLn (encode evt)
      hFlush stdout
      return Nothing
    Just (ParseResult _ profile bucket (Just prefix)) -> do
      -- Signal ready to git-lfs
      BLC.putStrLn "{}"
      hFlush stdout
      return (Just LFSProcess
        { lfsBucket  = bucket
        , lfsPrefix  = prefix
        , lfsProfile = profile
        , lfsS3Ops   = ops
        })

-- ---------------------------------------------------------------------------
-- Upload
-- ---------------------------------------------------------------------------

lfsUpload :: LFSProcess -> Aeson.Object -> IO ()
lfsUpload lfs event = do
  let ops   = lfsS3Ops lfs
      oid   = lookupText "oid"  event
      path  = lookupText "path" event
      key   = lfsPrefix lfs ++ "/lfs/" ++ oid
  result <- try (do
    exists <- lfsS3ObjectExists ops key
    if exists
      then do
        let evt = object [ "event" .= ("complete" :: String), "oid" .= oid ]
        BLC.putStrLn (encode evt)
        hFlush stdout
      else do
        cb <- newProgressCallback oid
        lfsS3UploadFile ops path key (reportProgress cb)
        let evt = object [ "event" .= ("complete" :: String), "oid" .= oid ]
        BLC.putStrLn (encode evt)
        hFlush stdout
    ) :: IO (Either SomeException ())
  case result of
    Left e  -> writeErrorEvent oid (show e)
    Right _ -> return ()

-- ---------------------------------------------------------------------------
-- Download
-- ---------------------------------------------------------------------------

lfsDownload :: LFSProcess -> Aeson.Object -> IO ()
lfsDownload lfs event = do
  let ops     = lfsS3Ops lfs
      oid     = lookupText "oid" event
      key     = lfsPrefix lfs ++ "/lfs/" ++ oid
  tempDir <- makeAbsolute ".git/lfs/tmp"
  createDirectoryIfMissing True tempDir
  let destPath = tempDir ++ "/" ++ oid
  result <- try (do
    cb <- newProgressCallback oid
    lfsS3DownloadFile ops key destPath (reportProgress cb)
    let evt = object
          [ "event" .= ("complete" :: String)
          , "oid"   .= oid
          , "path"  .= destPath
          ]
    BLC.putStrLn (encode evt)
    hFlush stdout
    ) :: IO (Either SomeException ())
  case result of
    Left e  -> writeErrorEvent oid (show e)
    Right _ -> return ()

-- ---------------------------------------------------------------------------
-- Helpers
-- ---------------------------------------------------------------------------

-- | Look up a text field in an Aeson Object, returning "" on failure.
lookupText :: String -> Aeson.Object -> String
lookupText k obj =
  case KM.lookup (AKey.fromString k) obj of
    Just (String t) -> T.unpack t
    _               -> ""

-- ---------------------------------------------------------------------------
-- install
-- ---------------------------------------------------------------------------

lfsInstall :: IO ()
lfsInstall = do
  (code1, _, err1) <- readProcessWithExitCode "git"
    ["config", "--add", "lfs.customtransfer.git-lfs-s3.path", "git-lfs-s3"] ""
  case code1 of
    ExitFailure _ -> do
      hPutStrLn stderr err1
      hFlush stderr
      exitWith (ExitFailure 1)
    ExitSuccess -> return ()

  (code2, _, err2) <- readProcessWithExitCode "git"
    ["config", "--add", "lfs.standalonetransferagent", "git-lfs-s3"] ""
  case code2 of
    ExitFailure _ -> do
      hPutStrLn stderr err2
      hFlush stderr
      exitWith (ExitFailure 1)
    ExitSuccess -> return ()

  putStrLn "git-lfs-s3 installed"
  hFlush stdout

-- ---------------------------------------------------------------------------
-- main loop
-- ---------------------------------------------------------------------------

-- | The main event loop for the git-lfs-s3 custom transfer agent.
-- Takes an 'LFSS3Ops' so that tests can inject a fake.
lfsMain :: LFSS3Ops -> [String] -> IO ()
lfsMain ops args = do
  hSetBuffering stdin  LineBuffering
  hSetBuffering stdout LineBuffering

  case args of
    ["install"] -> do
      lfsInstall
      exitSuccess
    ["enable-debug"] -> do
      _ <- readProcessWithExitCode "git"
        ["config", "--add", "lfs.customtransfer.git-lfs-s3.args", "debug"] ""
      putStrLn "debug enabled"
      exitSuccess
    ["disable-debug"] -> do
      _ <- readProcessWithExitCode "git"
        ["config", "--unset", "lfs.customtransfer.git-lfs-s3.args"] ""
      putStrLn "debug disabled"
      exitSuccess
    [unknown] -> do
      putStrLn ("unknown command " ++ unknown)
      exitWith (ExitFailure 1)
    _ -> runLfsLoop ops Nothing

runLfsLoop :: LFSS3Ops -> Maybe LFSProcess -> IO ()
runLfsLoop ops mLfs = do
  line <- getLine
  case Aeson.decode (BLC.pack line) :: Maybe Aeson.Value of
    Nothing -> do
      hPutStrLn stderr ("failed to parse JSON: " ++ line)
      runLfsLoop ops mLfs
    Just (Aeson.Object obj) -> do
      let event = lookupText "event" obj
      case event of
        "init" -> do
          let remoteName = lookupText "remote" obj
          unless (validateRefName remoteName) $ do
            BLC.putStrLn "{}"
            hFlush stdout
            exitWith (ExitFailure 1)
          (code, out, errStr) <- readProcessWithExitCode "git"
            ["remote", "get-url", remoteName] ""
          case code of
            ExitFailure _ -> do
              let errEvt = object
                    [ "error" .= object
                        [ "code"    .= (2 :: Int)
                        , "message" .= ("cannot resolve remote \"" ++ remoteName ++ "\"")
                        ]
                    ]
              BLC.putStrLn (encode errEvt)
              hFlush stdout
              exitWith (ExitFailure 1)
            ExitSuccess -> do
              let s3uri = reverse . dropWhile (== '\n') . reverse $ out
              newLfs <- newLFSProcess ops s3uri
              runLfsLoop ops newLfs

        "upload" ->
          case mLfs of
            Nothing  -> runLfsLoop ops mLfs
            Just lfs -> do
              lfsUpload lfs obj
              runLfsLoop ops (Just lfs)

        "download" ->
          case mLfs of
            Nothing  -> runLfsLoop ops mLfs
            Just lfs -> do
              lfsDownload lfs obj
              runLfsLoop ops (Just lfs)

        _ -> runLfsLoop ops mLfs

    _ -> runLfsLoop ops mLfs

File: src/GitRemoteS3/Lfs.php
<?php

// SPDX-FileCopyrightText: 2023-present Amazon.com, Inc. or its affiliates
//
// SPDX-License-Identifier: Apache-2.0

declare(strict_types=1);

namespace GitRemoteS3;

use Aws\S3\S3Client;
use Aws\S3\Exception\S3Exception;
use Monolog\Logger;
use Monolog\Handler\StreamHandler;

/**
 * Git LFS custom-transfer agent for Amazon S3.
 *
 * Corresponds to git_remote_s3/lfs.py.
 *
 * Protocol reference:
 *   https://github.com/git-lfs/git-lfs/blob/main/docs/custom-transfers.md
 *
 * NOTE: The Python implementation uses a threading.Lock for progress callbacks.
 *       PHP does not have native threads in the standard SAPI; progress events
 *       are therefore emitted without locking (safe in a single-threaded CLI).
 */
class Lfs
{
    private string   $prefix;
    private string   $bucket;
    private ?string  $profile;
    private S3Client $s3;

    private Logger $logger;

    public function __construct(string $s3Uri)
    {
        $this->logger = new Logger('git-lfs-s3');
        // Log to file — same path as Python implementation
        $logFile = '.git/lfs/tmp/git-lfs-s3.log';
        if (!is_dir(dirname($logFile))) {
            @mkdir(dirname($logFile), 0755, true);
        }
        try {
            $this->logger->pushHandler(new StreamHandler($logFile, Logger::ERROR));
        } catch (\Exception $e) {
            // If we cannot open the log file, silently continue
        }

        [$uriScheme, $profile, $bucket, $prefix] = Common::parseGitUrl($s3Uri);

        if ($bucket === null || $prefix === null) {
            $this->logger->error("s3 uri {$s3Uri} is invalid");
            $errorEvent = ['error' => ['code' => 32, 'message' => "s3 uri {$s3Uri} is invalid"]];
            fwrite(STDOUT, json_encode($errorEvent) . "\n");
            fflush(STDOUT);
            return;
        }

        $this->prefix  = $prefix;
        $this->bucket  = $bucket;
        $this->profile = $profile;

        // Send the initialisation acknowledgement expected by git-lfs
        fwrite(STDOUT, "{}\n");
        fflush(STDOUT);

        // Build S3 client lazily — done inline so tests can subclass
        $this->initS3Client();
    }

    private function initS3Client(): void
    {
        $region = getenv('AWS_DEFAULT_REGION') ?: (getenv('AWS_REGION') ?: 'us-east-1');
        $args = ['version' => 'latest', 'region' => $region];
        if ($this->profile !== null) {
            $args['profile'] = $this->profile;
        }
        $this->s3 = new S3Client($args);
    }

    // -------------------------------------------------------------------------
    // Upload / Download
    // -------------------------------------------------------------------------

    /**
     * Handle an LFS "upload" event.
     *
     * @param array{oid: string, path: string, size?: int} $event
     */
    public function upload(array $event): void
    {
        $this->logger->debug('upload');
        $oid = $event['oid'];
        try {
            // Check if object already exists
            $existing = $this->s3->listObjectsV2([
                'Bucket' => $this->bucket,
                'Prefix' => "{$this->prefix}/lfs/{$oid}",
            ])->get('Contents') ?? [];

            if (!empty($existing)) {
                $this->logger->debug('object already exists');
                fwrite(STDOUT, json_encode(['event' => 'complete', 'oid' => $oid]) . "\n");
                fflush(STDOUT);
                return;
            }

            // Upload with progress
            $bytesSoFar = 0;
            $this->s3->putObject([
                'Bucket'         => $this->bucket,
                'Key'            => "{$this->prefix}/lfs/{$oid}",
                'SourceFile'     => $event['path'],
                '@http'          => [
                    'progress' => function (
                        int $downloadTotal,
                        int $downloaded,
                        int $uploadTotal,
                        int $uploaded
                    ) use ($oid, &$bytesSoFar): void {
                        if ($uploaded > 0) {
                            $bytesSinceLast = $uploaded - $bytesSoFar;
                            if ($bytesSinceLast > 0) {
                                $bytesSoFar = $uploaded;
                                $progressEvent = [
                                    'event'          => 'progress',
                                    'oid'            => $oid,
                                    'bytesSoFar'     => $bytesSoFar,
                                    'bytesSinceLast' => $bytesSinceLast,
                                ];
                                fwrite(STDOUT, json_encode($progressEvent) . "\n");
                                fflush(STDOUT);
                            }
                        }
                    },
                ],
            ]);

            fwrite(STDOUT, json_encode(['event' => 'complete', 'oid' => $oid]) . "\n");
        } catch (\Exception $e) {
            $this->logger->error($e->getMessage());
            self::writeErrorEvent($oid, $e->getMessage());
        }
        fflush(STDOUT);
    }

    /**
     * Handle an LFS "download" event.
     *
     * @param array{oid: string} $event
     */
    public function download(array $event): void
    {
        $this->logger->debug('download');
        $oid     = $event['oid'];
        $tempDir = realpath('.git/lfs/tmp') ?: sys_get_temp_dir();
        $dest    = "{$tempDir}/{$oid}";

        try {
            $bytesSoFar = 0;
            $this->s3->getObject([
                'Bucket'  => $this->bucket,
                'Key'     => "{$this->prefix}/lfs/{$oid}",
                'SaveAs'  => $dest,
                '@http'   => [
                    'progress' => function (
                        int $downloadTotal,
                        int $downloaded,
                        int $uploadTotal,
                        int $uploaded
                    ) use ($oid, &$bytesSoFar): void {
                        if ($downloaded > 0) {
                            $bytesSinceLast = $downloaded - $bytesSoFar;
                            if ($bytesSinceLast > 0) {
                                $bytesSoFar = $downloaded;
                                $progressEvent = [
                                    'event'          => 'progress',
                                    'oid'            => $oid,
                                    'bytesSoFar'     => $bytesSoFar,
                                    'bytesSinceLast' => $bytesSinceLast,
                                ];
                                fwrite(STDOUT, json_encode($progressEvent) . "\n");
                                fflush(STDOUT);
                            }
                        }
                    },
                ],
            ]);

            $doneEvent = ['event' => 'complete', 'oid' => $oid, 'path' => $dest];
            fwrite(STDOUT, json_encode($doneEvent) . "\n");
        } catch (\Exception $e) {
            $this->logger->error($e->getMessage());
            self::writeErrorEvent($oid, $e->getMessage());
        }
        fflush(STDOUT);
    }

    // -------------------------------------------------------------------------
    // Static helpers
    // -------------------------------------------------------------------------

    public static function writeErrorEvent(string $oid, string $error, bool $flush = false): void
    {
        $errEvent = [
            'event' => 'complete',
            'oid'   => $oid,
            'error' => ['code' => 2, 'message' => $error],
        ];
        fwrite(STDOUT, json_encode($errEvent) . "\n");
        if ($flush) {
            fflush(STDOUT);
        }
    }

    /**
     * Install git-lfs-s3 as the custom transfer agent in the current repo.
     */
    public static function install(): void
    {
        exec(
            'git config --add lfs.customtransfer.git-lfs-s3.path git-lfs-s3',
            $output,
            $code
        );
        if ($code !== 0) {
            fwrite(STDERR, implode("\n", $output));
            fflush(STDERR);
            exit(1);
        }

        exec(
            'git config --add lfs.standalonetransferagent git-lfs-s3',
            $output,
            $code
        );
        if ($code !== 0) {
            fwrite(STDERR, implode("\n", $output));
            fflush(STDERR);
            exit(1);
        }

        fwrite(STDOUT, "git-lfs-s3 installed\n");
        fflush(STDOUT);
    }
}

// ---------------------------------------------------------------------------
// Entry-point
// ---------------------------------------------------------------------------

/**
 * Main function for the git-lfs-s3 custom transfer agent.
 *
 * Corresponds to git_remote_s3.lfs:main().
 */
function lfsMain(): void
{
    global $argv;

    if (isset($argv[1])) {
        switch ($argv[1]) {
            case 'install':
                Lfs::install();
                exit(0);

            case 'debug':
                // Enable debug logging — handled when the Lfs object is constructed.
                // Fall through to main loop with debug flag set via env or config.
                break;

            case 'enable-debug':
                exec('git config --add lfs.customtransfer.git-lfs-s3.args debug');
                echo "debug enabled\n";
                exit(0);

            case 'disable-debug':
                exec('git config --unset lfs.customtransfer.git-lfs-s3.args');
                echo "debug disabled\n";
                exit(0);

            default:
                echo "unknown command {$argv[1]}\n";
                exit(1);
        }
    }

    $lfsProcess = null;

    while (true) {
        $line = fgets(STDIN);
        if ($line === false) {
            break;
        }
        $event = json_decode($line, true);
        if (!is_array($event)) {
            continue;
        }

        switch ($event['event'] ?? '') {
            case 'init':
                $remote = $event['remote'] ?? '';
                if (!Git::validateRefName($remote)) {
                    fwrite(STDOUT, "{}\n");
                    fflush(STDOUT);
                    exit(1);
                }

                $s3Uri = '';
                exec('git remote get-url ' . escapeshellarg($remote), $output, $code);
                if ($code !== 0) {
                    $errEvent = [
                        'error' => [
                            'code'    => 2,
                            'message' => "cannot resolve remote \"{$remote}\"",
                        ],
                    ];
                    fwrite(STDOUT, json_encode($errEvent));
                    fflush(STDOUT);
                    exit(1);
                }
                $s3Uri = trim(implode('', $output));
                $lfsProcess = new Lfs($s3Uri);
                break;

            case 'upload':
                if ($lfsProcess !== null) {
                    $lfsProcess->upload($event);
                }
                break;

            case 'download':
                if ($lfsProcess !== null) {
                    $lfsProcess->download($event);
                }
                break;
        }
    }
}
